src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -0,0 +1,322 @@
+package org.broadinstitute.hellbender.tools.spark.transforms.markduplicates;
+
+import com.google.common.collect.*;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.metrics.MetricsFile;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.tools.dataflow.transforms.markduplicates.PairedEnds;
+import org.broadinstitute.hellbender.tools.dataflow.transforms.markduplicates.ReadsKey;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+import org.broadinstitute.hellbender.utils.read.markduplicates.DuplicationMetrics;
+import org.broadinstitute.hellbender.utils.read.markduplicates.LibraryIdGenerator;
+import org.broadinstitute.hellbender.utils.read.markduplicates.OpticalDuplicateFinder;
+import scala.Tuple2;
+
+import java.io.File;
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Utility classes and functions for Mark Duplicates.
+ */
+public class MarkDuplicatesSparkUtils {
+    //Bases below this quality will not be included in picking the best read from a set of duplicates.
+    public static final int MIN_BASE_QUAL = 15;
+
+    // Used to set an attribute on the GATKRead marking this read as an optical duplicate.
+    private static final String OPTICAL_DUPLICATE_TOTAL_ATTRIBUTE_NAME = "OD";
+
+    /**
+     * Takes the reads,
+     * group them by library, contig, position and orientation,
+     * within each group
+     *   (a) if there are only fragments, mark all but the highest scoring as duplicates, or,
+     *   (b) if at least one is marked as paired, mark all fragments as duplicates.
+     *  Note: Emit only the fragments, as the paired reads are handled separately.
+     */
+    static JavaRDD<GATKRead> transformFragments(final SAMFileHeader header, final JavaRDD<GATKRead> fragments) {
+        //
+        // Groups reads by keys - keys are tuples of (library, contig, position, orientation).
+        //
+
+        JavaPairRDD<String, Iterable<GATKRead>> groupedReads = fragments.mapToPair(read -> {
```

Comment:
This wouldn't accommodate the `read.setIsDuplicate(false)`. That could be done in a map just before `keyBy` though.